### PR TITLE
Increase support for authorization of Wechat Webpage

### PR DESCRIPTION
**Pull request by saber-wang**
增加对微信网页授权的支持
https://mp.weixin.qq.com/wiki?t=resource/res_main&id=mp1421140842

### CODE REVIEW

**Comment by martincostello** [comment]
Hi @saber-wang, thanks for your PR.

Could you explain what this PR is trying to achieving in English from a functional perspective please?

Unfortunately the maintainers here do not speak Cantonese/Mandarin, so the PR description and code comment do not help explain.

**Comment by saber-wang** [comment]
Resolve the problem of state up to 128 bytes in the authorization of Web pages on the Wechat client (authorization address https://open.weixin.qq.com/connect/oauth2/authorize), and add the necessary “# wechat_redirect” parameter，Look at #266

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs:13]: Please sort the usings as per the existing code style with `System.*` namespaces before others.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: New line between usings and namespace please.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: This seems like something we should not be depending on - what are you using that's _"pub-internal"_?

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Use `string.Equals(a, b, StringComparison.Ordinal)`.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Braces please.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Why `c`?

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: `oauthstate` should be a constant, which is then used in this method and the other.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Use `Use string.Equals(a, b, StringComparison.Ordinal)`. In fact, as it's being done twice, it could be factored out into a method:

```csharp
private bool IsWeixinAuthorizationEndpointInUse()
{
    return string.Equals(Options.AuthorizationEndpoint, WeixinAuthenticationDefaults.AuthorizationEndpoint, StringComparison.Ordinal);
}
```

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Relying on the name of a local variable for a parameter name is brittle against future refactoring. Use a constant.

Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Is this safe against URI encoding issues?

Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: You could simplify this by just checking whether you need to use `?` or `&`, and then just appending the state key and value.

**Comment by saber-wang** [comment]
Thank you for your comments.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (saber-wang) [reply src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Suddenly understand this problem, what good suggestions do you have?

Comment (martincostello) [reply src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: You could use [`Uri.EscapeDataString(string)`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.escapedatastring?view=netcore-2.2) or [`Uri.EscapeUriString(string)`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.escapeuristring?view=netcore-2.2) (I forget which is the right one).

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: Use `TryGetValue()` - it's more efficent that `ContainsKey()` and `[]`.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: So just to be clear, Weixin expects the `state` parameter value to always end with `#wechat_redirect`, or is what you're meaning to do is put a hash at the end of the whole URL?

I suspect it's the latter, otherwise you'd have to URI encode the `#`.

For that, you'd just do this:

```csharp
return QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, new Dictionary<string, string>
            {
                ["appid"] = Options.ClientId,
                ["scope"] = FormatScope(),
                ["response_type"] = "code",
                ["redirect_uri"] = redirectUri,
                ["state"] = oauthstate
                ["state"] = state
            }) + "#wechat_redirect";
```

Which could then probably be refactored to:

```csharp
var state = Options.StateDataFormat.Protect(properties);
var addRedirectHash = false;

if (!IsWeixinAuthorizationEndpointInUse())
{
    redirectUri = QueryHelpers.AddQueryString(redirectUri, OauthState, state);
    addRedirectHash = true;
}

redirectUri = QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, new Dictionary<string, string>
            {
                ["appid"] = Options.ClientId,
                ["scope"] = FormatScope(),
                ["response_type"] = "code",
                ["redirect_uri"] = redirectUri,
                ["state"] = oauthstate
                ["state"] = state
            });

if (addRedirectHash)
{
    // The parameters necessary for Web Authorization of Wechat
    redirectUri += "#wechat_redirect";
}

return redirectUri;
```

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs:161]: New line between the methods please.

File: src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs
Comment (saber-wang) [reply src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs]: `if (Request.Query.TryGetValue(OauthState, out var state))
                {
                    var query = Request.Query.ToDictionary(c => c.Key, c => c.Value, StringComparer.OrdinalIgnoreCase);
                    if (query.TryGetValue("sata",out var _))
                    {
                        query["sata"] = state;
                        Request.QueryString = QueryString.Create(query);
                    }
                }`
So?

**Comment by martincostello** [comment]
This looks OK to me now.

As part of the preparation for ASP.NET Core 3.0 support, tests are being added to help make things easier to maintain and validate going forwards (see #292).

Once tests are merged into the dev branch, can you rebase onto dev and update the relevant tests (if necessary) to account for these changes?

**Comment by saber-wang** [comment]
Asp.NetCore 3.0 has not been studied yet, but I think I can try it.

**Comment by martincostello** [comment]
The tests target 2.2, not 3.0. They’re being done in advance of the 3.0 work in #280 as it involves a lot refactoring.

**Comment by martincostello** [comment]
#280 has been merged to `dev` now. If you rebase and update any tests as appropriate, we can look at merging this PR soon.

**Comment by saber-wang** [comment]
The only difference with WeixinTests is Authorization Endpoint, but I added a new folder. I don't know if it's appropriate.

File: test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj]: This item group can be deleted - the file is already catered for by line 11.

File: test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: This looks like you're undoing what the code was trying to achieve, which was stripping out all the query string parameters except those explicitly in use (`code` and `state`). If you need to retain `redirect_uri` then you just need to read it out into a local like on line 38 and then put it back like on line 42.

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: But this is not a general parameter. I put state in redirect_uri to solve the problem of too long state parameter.

Comment (martincostello) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: From what I can see in the diff, you're forwarding on everything in the original query string for the request to the OAuth provider back to the app under test.

The original code is only forwarding back the parameters that were needed, which are `code` and `state`. If you also need `redirect_uri`, then just add that as special case too.

If we were going to just reflect everything back, then the majority of the code in that method becomes redundant and could be simplified compared to what's there now in your change as all the special-case code would become defunct as it would be passed back anyway.

The reason to only pick-and-choose what's reflected back is to remove noise from provider-specific parameters in the call back to the test app as part of the loopback is just to make it easier to debug.

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: so?
                var builder = new UriBuilder(location);
                // Retain the parameters of redirect_uri
                queryString = HttpUtility.ParseQueryString(builder.Query);
                var oauthstate= queryString["_oauthstate"];
                queryString.Clear();

                queryString.Add("code", "a6ed8e7f-471f-44f1-903b-65946475f351");
                queryString.Add("state", state);
                queryString.Add("_oauthstate", oauthstate);
                builder.Query = queryString.ToString();

Comment (martincostello) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: So like the below, you should only need to add two lines of code.

```csharp
string state = queryString["state"];
string oauthState = queryString["_oauthstate"];

queryString.Clear();

queryString.Add("code", "a6ed8e7f-471f-44f1-903b-65946475f351");
queryString.Add("state", state);
queryString.Add("_oauthstate", oauthState);

var builder = new UriBuilder(location)
{
    Query = queryString.ToString(),
};
```

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: _ oauthstate is in redirect_uri, so it needs to be parsed from location-_-

Comment (martincostello) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: Right I see, so then you only need to pass back `redirect_uri` then, not `_oauthstate`?

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: Yes

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: You need to take `_oauthstate `from `redirect_uri `and resolve it to `state`
Look at the code
```
protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
        {
            if (!IsWeixinAuthorizationEndpointInUse())
            {
                if (Request.Query.TryGetValue(OauthState, out var stateValue))
                {
                    var query = Request.Query.ToDictionary(c => c.Key, c => c.Value, StringComparer.OrdinalIgnoreCase);
                    if (query.TryGetValue(State, out var _))
                    {
                        query[State] = stateValue;
                        Request.QueryString = QueryString.Create(query);
                    }
                }
            }
            return await base.HandleRemoteAuthenticateAsync();
        }
```

Comment (martincostello) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: So for the tests you just need `redirect_uri` passed back then.

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs]: `code` is also required, so please check my submission

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: code is also required, so please check my submission

Comment (saber-wang) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: So what should I do now?

Comment (martincostello) [reply test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs:37]: I'll have time to re-review this later today, or at the latest this weekend.

**Review by martincostello — APPROVED** [review]

**Comment by saber-wang** [comment]
@martincostello  Thank you for helping me revise these.

**Comment by liminany** [comment]
@saber-wang 你好，目前我使用2.1版本测试情况如下：
当只配置一个Weixin Provider时（开放平台和公众平台二选一）时没有问题

但是，我同时配置两个Weixin Provider的时候不行了。Startpup.cs里面的配置如下：

image


请问，如果 要同时支持PC网页扫描登陆和微信客户端登陆，需要怎么样配置？

从我的测试来看，目前是不支持同时配置两个端。期待你的回复，谢谢！

**Comment by saber-wang** [comment]
@liminany 在asp.net core中实际是通过CallbackPath来匹配Scheme的，所以你只需要配置不同的CallbackPath和Scheme就可以了例如`options.CallbackPath = "/signin-Wechat"`

**Comment by liminany** [comment]
@saber-wang 非常感谢，没找到对应的使用文档，这个问题困扰了我好几天了，配置好callbackPath就可以了。

另外 在提个问题：如果还想要在手机浏览器（比如UC浏览器）中实现微信登陆应该怎样做？

**Comment by saber-wang** [comment]
@liminany 在微信外部正常情况下都只能通过扫码登陆

**Comment by liminany** [comment]
@saber-wang PC端登陆可以，但换到手机浏览器，长按二维码登陆，提示 Scope参数错误或没有Scope权限 ，不知道什么原因

**Comment by saber-wang** [comment]
@liminany Scope参数错误或没有Scope权限 这个问题我也遇到了，但是是在PC扫码就提示这个，目前还没有找到问题所在

**Comment by StarkShang** [comment]
@saber-wang 微信登陆回调地址是不是可以携带额外参数？
return Challenge(new AuthenticationProperties { RedirectUri = redirectUrl }, dto.Provider);

**Comment by saber-wang** [comment]
@StarkShang Splice it into the redirecturl

**Comment by StarkShang** [comment]
@saber-wang redirecturi is configured in startup.cs, and changing the redirectUrl here has no effect. There is no query parameters in returnUrl.

**Comment by saber-wang** [comment]
@StarkShang return Challenge(new AuthenticationProperties { RedirectUri = **redirectUrl** }, dto.Provider); here

**Pull request merged**
